cmd/collect_signals/localworker/loop.go

```diff
@@ -0,0 +1,232 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package localworker
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/ossf/scorecard/v4/cron/config"
+	"github.com/ossf/scorecard/v4/cron/data"
+	"github.com/ossf/scorecard/v4/cron/worker"
+	"go.uber.org/zap"
+
+	"github.com/ossf/criticality_score/internal/iterator"
+)
+
+// maxAttempts is the maximum number of times a batch will be attempted before
+// giving up and moving on to the next batch.
+//
+// TODO: make this configurable.
+const maxAttempts = 7
+
+const (
+	configLocalURLDataFile = "local-url-data-file"
+	configLocalStateFile   = "local-state-file"
+)
+
+// WorkLoop is a version of worker.WorkLoop that reads data from a local file, rather
+// than a from a pubsub feed populated by the Scorecard controller.
+type WorkLoop struct {
+	logger *zap.Logger
+	w      worker.Worker
+	input  iterator.IterCloser[string]
+
+	// StateFilename points to the file used to store progress to allow for
+	// recovery if the worker is terminated for any reason. The name should not
+	// change for a given input.
+	stateFilename string
+
+	// BucketURL is a gocloud.dev URL pointing to the data bucket where data
+	// will be written.
+	bucketURL string
+
+	// RawBucketURL is a gocloud.dev URL pointing to the raw data bucket where
+	// raw data will be written.
+	rawBucketURL string
+
+	// ShardSize defines the size of each shard passed to the worker for
+	// processing.
+	shardSize int
+}
+
+// FromConfig returns a new local WorkLoop instance using the config values
+// read in.
+//
+// Apart from the usual config values for the data bucket, raw data bucket, and
+// shard size, the criticality score config values "local-url-data-file" and
+// "local-state-file" are also used.
+func FromConfig(logger *zap.Logger, w worker.Worker) (*WorkLoop, error) {
+	if !flag.Parsed() {
+		flag.Parse()
+	}
+
+	if err := config.ReadConfig(); err != nil {
+		return nil, fmt.Errorf("reading config: %w", err)
+	}
+
+	bucketURL, err := config.GetResultDataBucketURL()
+	if err != nil {
+		return nil, fmt.Errorf("config data bucket url: %w", err)
+	}
+
+	rawBucketURL, err := config.GetRawResultDataBucketURL()
+	if err != nil {
+		return nil, fmt.Errorf("config raw data bucket url: %w", err)
+	}
+	shardSize, err := config.GetShardSize()
+	if err != nil {
+		return nil, fmt.Errorf("config shard size: %w", err)
+	}
+
+	criticalityConfig, err := config.GetCriticalityValues()
+	if err != nil {
+		return nil, fmt.Errorf("criticality score config: %w", err)
+	}
+
+	inFile := criticalityConfig[configLocalURLDataFile]
+	if inFile == "" {
+		return nil, fmt.Errorf("%q config not set", configLocalURLDataFile)
+	}
+	f, err := os.Open(inFile)
```

Comment:
Who closes this?

Reply:
Oh thanks! I forgot to close the iterator which closes the file.